x-pack/platform/packages/shared/onechat/onechat-server/tools/builtin.ts

```diff
@@ -5,13 +5,69 @@
  * 2.0.
  */
 
+import type { MaybePromise } from '@kbn/utility-types';
 import type { z, ZodObject } from '@kbn/zod';
+import type { IUiSettingsClient } from '@kbn/core-ui-settings-server';
 import type { ToolDefinition, ToolType } from '@kbn/onechat-common';
 import type { EsqlToolDefinition } from '@kbn/onechat-common/tools/types/esql';
 import type { IndexSearchToolDefinition } from '@kbn/onechat-common/tools/types/index_search';
 import type { WorkflowToolDefinition } from '@kbn/onechat-common/tools/types/workflow';
+import type { KibanaRequest } from '@kbn/core-http-server';
 import type { ToolHandlerFn } from './handler';
 
+/**
+ * Information exposed to the {@link ToolAvailabilityHandler}.
+ */
+export interface ToolAvailabilityContext {
+  request: KibanaRequest;
+  uiSettings: IUiSettingsClient;
+  spaceId: string;
+}
+
+/**
+ * Information exposed to the {@link ToolAvailabilityHandler}.
+ */
+export interface ToolAvailabilityResult {
+  /**
+   * Whether the tool is available or not.
+   */
+  status: 'available' | 'unavailable';
+  /**
+   * Optional reason for why the tool is unavailable.
+   */
+  reason?: string;
+}
+
+/**
+ * Availability handler for a tool.
+ */
+export type ToolAvailabilityHandler = (
+  context: ToolAvailabilityContext
+) => MaybePromise<ToolAvailabilityResult>;
+
+export interface ToolAvailabilityConfig {
+  /**
+   * handler which can be defined to add conditional availability of the tool.
+   *
+   * Note: this is meant to be used for tools that are gated behind a feature flag,
+   *       or tools which have some condition to be available.
+   *       it *IS NOT* meant to be used as a replacement for RBAC.
+   */
+  handler: ToolAvailabilityHandler;
+  /**
+   * Cache mode for the result
+   * - global: the result will be cached globally, for all spaces
+   * - space: the result will be cached per-space
+   * - none: the result shouldn't be cached (warning: this can lead to performance issues)
+   */
+  cacheMode: 'global' | 'space' | 'none';
```

Comment:
I was scared of the performance aspect of calling that handler for each tool implementing it, every time we need to list or retrieve tools, so I added built-in cache support for it. It's not perfect, but it should significantly reduces the risk of perf abuse.